### ISSUE TITLE
[REQ] Alternative "discovery agent" for non-bluetooth devices

### ISSUE THREAD

**Issue by drmason789** [cagnulein/qdomyos-zwift#889]
As I currently understand it, non-bluetooth devices cannot be reliably detected by placing their detection code in bluetooth::deviceDiscovered(...) along with the others, because that code is only called if a Bluetooth device is found.

The solution I have is to put the detection code in the bluetooth class constructor. This might only produce a tiny delay when creating the object, but as more such device detectors are added, this could become noticable.

Perhaps there could be an alternative to the QBluetoothDeviceDiscoveryAgent which runs in a different thread, and contains the code for detecting non-Bluetooth devices. 

It should be clearly specified what a detector function needs to deliver, i.e. an object of a class inheriting from bluetoothdevice, and other things, e.g. a QBluetoothDeviceInfo object with a UUID generated for the device type (because it doesn't have one) and name to display in the UI.

**Comment by cagnulein** [comment]
I don't know if I get your idea, so I will try to explain mine and then we can merge our ideas :)

The actual Bluetooth constructor manages all the Bluetooth devices. Done.
Inside of the constructors I would like to add 3 other threads:

1. Wi-Fi/Ethernet thread handler (for bike/treadmill like the new Nordictrack or proform...)
2. Serial port (like your bike)
3. Gpio handler (for the gpio treadmill branch)

Maybe I'm forgetting some other cases, but I don't think that they should be more than 5, so I would not worry about latency since the constructor will be called one time at the startup of QZ

In each thread, every class could be add its handler to catch the logic to detect the device (so, for example, for your bike, listening the serial ports..)

What do you think?

**Comment by drmason789** [comment]
@cagnulein I believe I derived my idea from your idea in #855 . I agree with your details. I'm willing to do this for the serial port.

**Comment by drmason789** [comment]
@cagnulein I'm looking at aligning the starting and stopping of discovery of non-bluetooth devices with the that of the bluetooth device discovery, and I see that the bluetooth discovery is started in different ways on different platforms, but this is not consistent across all "starts".

For example,  bluetooth::bluetooth and bluetooth::finished have:

```
    if (!trx_route_key && !bh_spada_2 && !technogym_myrun_treadmill_experimental) {
#endif
        discoveryAgent->start(QBluetoothDeviceDiscoveryAgent::LowEnergyMethod);
#ifndef Q_OS_IOS
    } else {
        discoveryAgent->start(QBluetoothDeviceDiscoveryAgent::ClassicMethod |
                              QBluetoothDeviceDiscoveryAgent::LowEnergyMethod);
    }
#endif
```



but bluetooth::restart has:

```
    if (onlyDiscover) {
        onlyDiscover = false;
        discoveryAgent->start();
        return;
    }
```

and at the end

`discoveryAgent->start();`


Should these actually all be the same?

**Comment by drmason789** [comment]
I've committed a rudimentary form of this in my branch for #855 but I can do something I consider to be much better for this issue. For the above inconsistency, I've factored out the various discoveryAgent->start(...) into a method with a bool parameter to indicate if it should supply discovery mode information.

**Comment by cagnulein** [comment]
> Should these actually all be the same?

yes you're right

**Comment by drmason789** [comment]
Note to self: cancel the discovery process on both the real-bluetooth device and non-bluetooth device threads once a device has been found.

**Comment by drmason789** [comment]
I'm waiting for a decision on #910 before proceeding with this.

**Comment by drmason789** [comment]
I'm thinking that the device detection functionality should be in a static function in the device's class itself, at least, to hide the details from the bluetooth discovery code that is currently implemented as a very long if else if else statement, which I converted to functions in my first PR for #887. 

These static functions can be registered in a vector and looped through in the discovery process.

This could be extended so that each detector is a subclass of an abstract "detector" class which provides members to determine separately, 
- if the device should be looked for
- detect the device
- create the bluetoothdevice object
- provide a device name for the UI for the "device name found" message.

**Comment by cagnulein** [comment]
Sounds nice. Probably moving all the fitness machine files in a subdir and looping it with a macro will also remove the needs of populate the static array of each static function for the fitness machine modules

Anyway it could be a huge work and maybe it's better adding the test cases first in order to be sure to don't destroy anything :)

**Comment by drmason789** [comment]
Yes, I agree about the test cases (#911). I'm just making notes, thinking ahead.

**Comment by drmason789** [comment]
Related to #927

**Comment by drmason789** [comment]
Regarding the "wontfix" label, I haven't forgotten about it. I am just assigning a higher priority to these: #911 then #887

**Comment by cagnulein** [comment]
The won't fix label was put from the bot :)
I removed with the pinned state :)
